flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/FileRegionBuffer.java

```diff
@@ -0,0 +1,213 @@
+
+import org.apache.flink.shaded.netty4.io.netty.buffer.ByteBuf;
+import org.apache.flink.shaded.netty4.io.netty.buffer.ByteBufAllocator;
+import org.apache.flink.shaded.netty4.io.netty.channel.DefaultFileRegion;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Wrapper for {@link org.apache.flink.shaded.netty4.io.netty.channel.FileRegion} instance.
+ *
+ * <p>The file region can be sent via network channel which supports zero-copy file transfer,
+ * and it can also be read into memory segment locally via {@link #readInto(MemorySegment)}.
+ */
+public class FileRegionBuffer extends DefaultFileRegion implements Buffer {
+
+	/** The number of bytes to be read/transferred from this file region. */
+	private final int bufferSize;
+
+	private final FileChannel fileChannel;
+
+	/** The {@link DataType} this buffer represents. */
+	private final DataType dataType;
+
+	/** Whether the buffer is compressed or not. */
+	private final boolean isCompressed;
+
+	public FileRegionBuffer(
+			FileChannel fileChannel,
+			int bufferSize,
+			DataType dataType,
+			boolean isCompressed) throws IOException {
+
+		super(fileChannel, fileChannel.position(), bufferSize);
+
+		this.fileChannel = checkNotNull(fileChannel);
+		this.bufferSize = bufferSize;
+		this.dataType = checkNotNull(dataType);
+		this.isCompressed = isCompressed;
+	}
+
+	// ------------------------------------------------------------------------
+	// Buffer override methods
+	// ------------------------------------------------------------------------
+
+	@Override
+	public boolean isBuffer() {
+		return dataType.isBuffer();
+	}
+
+	@Override
+	public MemorySegment getMemorySegment() {
+		throw new UnsupportedOperationException("Method should never be called.");
+	}
+
+	@Override
+	public int getMemorySegmentOffset() {
+		throw new UnsupportedOperationException("Method should never be called.");
+	}
+
+	@Override
+	public ReadOnlySlicedNetworkBuffer readOnlySlice() {
+		throw new UnsupportedOperationException("Method should never be called.");
+	}
+
+	@Override
+	public ReadOnlySlicedNetworkBuffer readOnlySlice(int index, int length) {
+		throw new UnsupportedOperationException("Method should never be called.");
+	}
+
+	@Override
+	public int getMaxCapacity() {
+		throw new UnsupportedOperationException("Method should never be called.");
+	}
+
+	@Override
+	public int getReaderIndex() {
+		throw new UnsupportedOperationException("Method should never be called.");
+	}
+
+	@Override
+	public void setReaderIndex(int readerIndex) throws IndexOutOfBoundsException {
+		throw new UnsupportedOperationException("Method should never be called.");
+	}
+
+	/**
+	 * This method is only implemented for tests at the moment.
+	 */
+	@Override
+	public ByteBuffer getNioBufferReadable() {
+		return BufferReaderWriterUtil.tryReadByteBuffer(fileChannel, bufferSize);
```

Comment:
This method here is a little concerning, because if used wrong it voids the whole implementation. It reads into a newly allocated buffer.

I assume this is mainly needed for events?

One alternative here could be that the reading of events from the file channel produces a regular `NetworkBuffer`. Then this method here would not be needed.

Reply:
As I mentioned in javadoc for this method, it is only used for unit tests ATM, not for the core codes.
Actually in the first version I made this method for `UnsupportedOperationException`, but I found many exiting tests rely on this method like `BoundedBlockingSubpartitionWriteReadTest`. So I implemented this method for avoiding refactoring the logic of unit tests for less efforts. Do you think we should refactor the unit test to avoid calling this way instead?